src/device_memory_tracker.h

```diff
@@ -0,0 +1,265 @@
+  // [WIP] Document thread-local nature..
+  static void TrackThreadMemoryUsage(ScopedMemoryUsage* usage);
+  static void UntrackThreadMemoryUsage(ScopedMemoryUsage* usage);
+
+  static void TrackActivity(CUpti_Activity* record)
+  {
+    tracker_->TrackActivityInternal(record);
+  }
+
+ private:
+  void TrackActivityInternal(CUpti_Activity* record);
+
+  static std::unique_ptr<DeviceMemoryTracker> tracker_;
+  std::mutex mtx_;
+  std::unordered_map<uint32_t, uintptr_t> activity_to_memory_usage_;
+};
+
+std::unique_ptr<DeviceMemoryTracker> DeviceMemoryTracker::tracker_{nullptr};
+
+// Boilerplate from CUPTI examples
+namespace {
+
+#define CUPTI_CALL(call)                                      \
+  do {                                                        \
+    CUptiResult _status = call;                               \
+    if (_status != CUPTI_SUCCESS) {                           \
+      const char* errstr;                                     \
+      cuptiGetResultString(_status, &errstr);                 \
+      LOG_ERROR << #call << " failed with error: " << errstr; \
+      exit(EXIT_FAILURE);                                     \
+    }                                                         \
+  } while (0)
+
+#define BUF_SIZE (32 * 1024)
+#define ALIGN_SIZE (8)
+#define ALIGN_BUFFER(buffer, align)                                 \
+  (((uintptr_t)(buffer) & ((align)-1))                              \
+       ? ((buffer) + (align) - ((uintptr_t)(buffer) & ((align)-1))) \
+       : (buffer))
+
+void
+bufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords)
+{
+  uint8_t* bfr = (uint8_t*)malloc(BUF_SIZE + ALIGN_SIZE);
+  if (bfr != nullptr) {
+    *size = BUF_SIZE;
+    *buffer = ALIGN_BUFFER(bfr, ALIGN_SIZE);
+    *maxNumRecords = 0;
+  } else {
+    LOG_ERROR << "Failed to allocate buffer for CUPIT: out of memory";
+  }
+}
+
+void
+bufferCompleted(
+    CUcontext ctx, uint32_t streamId, uint8_t* buffer, size_t size,
+    size_t validSize)
+{
+  CUptiResult status;
+  CUpti_Activity* record = nullptr;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(buffer, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+        DeviceMemoryTracker::TrackActivity(record);
+      } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED)
+        break;
+      else {
+        CUPTI_CALL(status);
+      }
+    } while (1);
+
+    // report any records dropped from the queue
+    size_t dropped;
+    CUPTI_CALL(cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped));
+    if (dropped != 0) {
+      LOG_WARNING << "Dropped " << dropped << " activity records";
+    }
+  }
+
+  free(buffer);
+}
+
+}  // namespace
+
+void
+DeviceMemoryTracker::InitTrace()
+{
+  if (tracker_ == nullptr) {
+    tracker_.reset(new DeviceMemoryTracker());
+    CUPTI_CALL(
+        cuptiActivityRegisterCallbacks(bufferRequested, bufferCompleted));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
```

Comment:
Do we need to disable the activity to make sure it doesn't add overhead for actual inferences?

Reply:
I believe enable/disable tracing is a global property so switching that will interfere with other memory usage collection, I have done some small performance comparison with the feature enable and the performance impact is not obvious (on the simple test models). But I do agree that it will add overhead to inference, I would like to see the impact on small model but with large number of kernel launches (speed on launching CUDA activities is bottleneck), but I don't have such a model in hand.